File: app/src/main/java/fr/xgouchet/chronorg/data/repositories/JumpRepository.java
package fr.xgouchet.chronorg.data.repositories;

import android.content.ContentResolver;
import android.content.Context;
import android.support.annotation.NonNull;

import com.deezer.android.counsel.annotations.Trace;

import fr.xgouchet.chronorg.data.ioproviders.JumpIOProvider;
import fr.xgouchet.chronorg.data.models.Jump;
import fr.xgouchet.chronorg.data.queriers.JumpContentQuerier;
import rx.Observable;
import rx.Subscriber;

/**
 * @author Xavier Gouchet
 */
@Trace
public class JumpRepository extends BaseRepository<Jump> {

    public JumpRepository(@NonNull Context context,
                          @NonNull JumpIOProvider provider) {
        super(context, provider);
    }


    public Observable<Jump> getJumpsInEntity(final int entityId) {
        return Observable.create(new Observable.OnSubscribe<Jump>() {
            @Override public void call(final Subscriber<? super Jump> subscriber) {
                try {
                    ContentResolver contentResolver = context.getContentResolver();
                    JumpContentQuerier jumpContentQuerier = (JumpContentQuerier) provider.provideQuerier();

                    jumpContentQuerier.queryInEntity(contentResolver,
                            new SubscriberWrapperAction<>(subscriber), entityId);
                    subscriber.onCompleted();
                } catch (Exception err) {
                    subscriber.onError(err);
                }
            }
        });
    }


}

File: khronorg/src/main/java/fr/xgouchet/khronorg/feature/jumps/JumpRepository.kt
package fr.xgouchet.khronorg.feature.jumps

import android.content.Context
import fr.xgouchet.khronorg.commons.ioproviders.IOProvider
import fr.xgouchet.khronorg.commons.query.QueryBuilder
import fr.xgouchet.khronorg.commons.repositories.BaseRepository
import fr.xgouchet.khronorg.data.models.Direction
import fr.xgouchet.khronorg.feature.travellers.Traveller
import fr.xgouchet.khronorg.provider.KhronorgSchema
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.functions.Consumer
import org.joda.time.Duration
import org.joda.time.Interval

/**
 * @author Xavier F. Gouchet
 */
class JumpRepository(context: Context, provider: IOProvider<Jump>) : BaseRepository<Jump>(context, provider) {

    fun getTravellerJumps(traveller: Traveller): Observable<Jump> {
        val source = ObservableOnSubscribe<Jump> {
            emitter ->
            try {
                // Birth
                val birth = Jump(id = Jump.ID_BIRTH,
                        travellerId = traveller.id,
                        from = traveller.birth,
                        delay = Interval(traveller.birth, Duration.ZERO),
                        direction = Direction.FUTURE)
                emitter.onNext(birth)

                // User defined jumps
                val query = QueryBuilder.where { equals(KhronorgSchema.COL_TRAVELLER_ID, traveller.id.toString()) }
                val contentResolver = context.contentResolver
                provider.querier.queryWhere(contentResolver, query, Consumer { t -> emitter.onNext(t) })

                // Death
                val death = Jump(id = Jump.ID_DEATH,
                        travellerId = traveller.id,
                        from = traveller.death,
                        delay = Interval(traveller.death, Duration.ZERO),
                        direction = Direction.FUTURE)
                emitter.onNext(death)

                // complete
                emitter.onComplete()
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }

        return Observable.create(source)
    }

}